pyserini/search/_dsearcher.py

```diff
@@ -0,0 +1,198 @@
+        self.idx2id, self.docs = self.load_corpus(doc_path)
+
+    @classmethod
+    def from_prebuilt_index(cls, prebuilt_index_name: str, corpus_name: str = None):
+        """Build a searcher from a pre-built index; download the index if necessary.
+
+        Parameters
+        ----------
+        corpus_name : str
+            corpus name
+        prebuilt_index_name : str
+            Prebuilt index name.
+
+        Returns
+        -------
+        SimpleDenseSearcher
+            Searcher built from the prebuilt faiss index.
+        """
+        return cls(prebuilt_index_name, corpus_name)
+
+    def search(self, emb_q: np.array, k: int = 10) -> List[DenseSearchResult]:
+        """Search the collection.
+
+        Parameters
+        ----------
+        emb_q : np.array
+            query embedding
+        k : int
+            Number of hits to return.
+        Returns
+        -------
+        List[DenseSearchResult]
+            List of search results.
+        """
+        assert len(emb_q) == self.dimension
+        emb_q = emb_q.reshape((1, len(emb_q)))
+        distances, indexes = self.index.search(emb_q, k)
+        distances = distances.flat
+        indexes = indexes.flat
+        if self.idx2id is None:
+            return [DenseSearchResult(str(idx), score, None) for score, idx in zip(distances, indexes) if idx != -1]
+        else:
+            return [
+                DenseSearchResult(self.idx2id[idx], score, self.docs[self.idx2id[idx]])
+                for score, idx in zip(distances, indexes) if idx != -1
+            ]
+
+    def batch_search(self, q_embs: List[np.array], q_ids: List[str], k: int = 10, threads: int = 1) -> List[DenseSearchResult]:
+        """
+
+        Parameters
+        ----------
+        q_embs : List[np.array]
+            List of query embeddings
+        q_ids : List[str]
+            List of corresponding query ids.
+        k : int
+            Number of hits to return.
+        threads : int
+            Maximum number of threads to use.
+
+        Returns
+        -------
+        Dict[str, List[DenseSearchResult]]
+            Dictionary holding the search results, with the query ids as keys and the corresponding lists of search
+            results as the values.
+        """
+        n, m = q_embs.shape
+        assert m == self.dimension
+        faiss.omp_set_num_threads(threads)
+        D, I = self.index.search(q_embs, k)
+        if self.idx2id is None:
+            return {key: [DenseSearchResult(str(idx), score, None)
+                          for score, idx in zip(distances, indexes) if idx != -1]
+                    for key, distances, indexes in zip(q_ids, D, I)}
+        else:
+            return {key: [DenseSearchResult(self.idx2id[idx], score, self.docs[self.idx2id[idx]])
+                          for score, idx in zip(distances, indexes) if idx != -1]
+                    for key, distances, indexes in zip(q_ids, D, I)}
+
+    def doc(self, docid: str):
```

Comment:
Do you think it's necessary to have this? We can just depend on the sparse index (Lucene) if the user really wants the raw text?

Reply:
we could, but in that case, user has to build a anserini index first?

Reply:
Yes, but we'll have prebuilt sparse indexes for ever dense index we have.

Reply:
another thing is the dense search only gives the index (the position of the document), we need a way to obtain the map of idx&docid, even if we eventually use sparse index to get raw text. rn I am getting the idx_id map by raw corpus which contains the docid.

Reply:
Oh, I see. In that case let's keep it then.